### PR TITLE
feat(python): expose row id in python API

### PR DESCRIPTION
**Pull request by eddyxu**
Added and modified two APIs
* Dataset.scanner(with_row_id)
* Dataset._take_rows(row_ids)

### CODE REVIEW

File: python/python/lance/dataset.py [hunk 1]
```diff
@@ -120,6 +120,7 @@ def scanner(
         fragment_readahead: Optional[int] = None,
         scan_in_order: bool = True,
         fragments: Optional[Iterable[LanceFragment]] = None,
+        with_row_id: bool = False,
```
Comment (wjones127) [inline python/python/lance/dataset.py:123]: Could we move this after the `*`? I think we want to do that going forward with all kwargs.

File: python/python/lance/dataset.py [hunk 2]
```diff
@@ -175,6 +176,10 @@ def scanner(
             number of rows (or be empty) if the rows closest to the query do not
             match the filter.  It's generally good when the filter is not very
             selective.
+        with_row_id: bool, default False
+            If set true, a column named ``_row_id`` will be added to the result, which
+            represents the internal id of the each row.
+
```
Comment (wjones127) [inline python/python/lance/dataset.py:182]: This is internal use only right? Perhaps we shouldn't document this option publicly?

File: python/python/lance/dataset.py (hunk 1 above)
Comment (eddyxu) [reply python/python/lance/dataset.py:123]: Ok, sg.

File: python/python/lance/dataset.py (hunk 2 above)
Comment (eddyxu) [reply python/python/lance/dataset.py:182]: Ok, fair.

**Review by wjones127 — APPROVED** [review]
Looks good.

BTW, I'm thinking we might want to rename "row_ids" to "row_addresses". I think that better communicates what they are, and the fact that they can change over time.

**Comment by eddyxu** [comment]
> BTW, I'm thinking we might want to rename "row_ids" to "row_addresses". I think that better communicates what they are, and the fact that they can change over time.

This is similar to `oid` in postgres or `row_number` in BigQuery.

**Pull request merged**